### DISCUSSION TITLE
Higher-refresh LEDs such as APA10*?

### DISCUSSION THREAD

**Discussion by Joshfindit** [PlummersSoftwareLLC/NightDriverStrip#40] [General]
So much respect for your work (before and now).

The 800Hz PWM on the WS2812 distracting enough that I jumped ship to the APA102s as soon as I could get my hands on them, and I've used them (and their siblings) since.

_Let's be clear: I know that I'm an outlier on this._

Does Night Driver have built-in support for these? If no: what would it take?

**Comment by davepl** [comment]
Not yet, but if you can convince me there’s something cool I can do with APA102s that I can’t do with WS2812B, that’d be plenty of incentive for me to add support!

0- Dave

**Comment by Joshfindit** [comment]
Besides the higher PWM (which must be seen to be judged):

- The APAs use less microcontroller CPU as they have an embedded chip that's controlled with the SPI protocol (I've even set colour and then disconnected the controller completely and the colour stays as it was set)
- Because they use a clock line, the timing is more precise while also being robust going down the line
- With the WS2812B you have to be precise on the 800Hz coming out of the controller, with the APAs you do not (this opens the possibility of multitasking and using interrupts)
- The APA102C has a 5-bit colour-independent brightness control (_caveat: there might be some poor implementations of this where setting the brightness this way causes the "brightness PWM" to be 500-600Hz_)
- Speaking of brightness: the APAs are significantly better at low brightness: they have better colour accuracy, go lower, and are stable when at low brightness (even on film. It's practically impossible to catch 19.2 kHz PWM with 60 or even 120 fps)
- POV
- Faster/bigger 2D and 3D grids

**Comment by Joshfindit** [comment]
As well: FastLED already supports them, so there is battle-tested code available

**Comment by ohnoitsalobo** [comment]
https://github.com/FastLED/FastLED/wiki/Overview#spichipsets

APA102 is natively supported by FastLED. You'd just need to adjust the parameters when initializing the LED strip(s) in the setup function. It should look something like this (please verify for yourself). 

```
FastLED.addLeds<APA102, CLK_PIN, DATA_PIN>(leds, 60);                 
```
You shouldn't need to adjust anything about NightDriver, just make sure the wires are connected to the correct pins.

**Reply by Joshfindit** [reply]
Thanks @ohnoitsalobo. I finally got a weekend to try this out. changed the line in `NUM_CHANNELS ==1` to be:
`FastLED.addLeds<APA102, GPIO16, GPIO17, COLOR_ORDER>(leds, g_pStrands[0]->GetLEDCount());`

But when uploading the `demo` and running it, the LEDs did not update.
